src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,79 @@
+        /// </summary>
+        public ImmutableArray<double> PerClassPrecision { get; }
+
+        /// <summary>
+        /// The calculated value of <a href="https://en.wikipedia.org/wiki/Precision_and_recall#Recall">recall</a> for each class.
+        /// </summary>
+        public ImmutableArray<double> PerClassRecall { get; }
+
+        /// <summary>
+        /// The confsion matrix counts for the combinations actual class/predicted class.
+        /// The actual classes are in the rows of the table, and the predicted classes in the columns.
+        /// </summary>
+        public ImmutableArray<ImmutableArray<double>> ConfusionTableCounts { get; }
+
+        /// <summary>
+        /// The indicators of the predicted classes.
+        /// It might be the classes names, or just indices of the predicted classes, if the name mapping is missing.
+        /// </summary>
+        public IReadOnlyList<ReadOnlyMemory<char>> PredictedClassesIndicators;
+
+        internal readonly bool Sampled;
+        internal readonly bool Binary;
+
+        private readonly IHost _host;
+
+        internal ConfusionMatrix(IHost host, double[] precision, double[] recall, double[][] confusionTableCounts,
+            List<ReadOnlyMemory<char>> labelNames, bool sampled, bool binary, DataViewSchema.Annotations classIndicators)
+        {
+            _host = host;
+
+            PerClassPrecision = precision.ToImmutableArray();
+            PerClassRecall = recall.ToImmutableArray();
+            Sampled = sampled;
+            Binary = binary;
+            PredictedClassesIndicators = labelNames.AsReadOnly();
+
+            var classNumber = confusionTableCounts.Length;
+            List<ImmutableArray<double>> counts = new List<ImmutableArray<double>>(classNumber);
+
+            for (int i = 0; i < classNumber; i++)
+                counts.Add(ImmutableArray.Create(confusionTableCounts[i]));
+
+            ConfusionTableCounts = counts.ToImmutableArray();
+
+        }
+
+        /// <summary>
+        /// Returns a human readable representation of the confusion table.
+        /// </summary>
+        /// <returns></returns>
+        public string GetFormattedConfusionTable() => MetricWriter.GetConfusionTableAsString(this, false);
+
+        /// <summary>
+        /// Gets the confusion table count for the pair <paramref name="predictedClassIndicatorIndex"/>/<paramref name="actualClassIndicatorIndex"/>.
+        /// </summary>
+        /// <param name="predictedClassIndicatorIndex">The index of the predicted label indicator, in the <see cref="PredictedClassesIndicators"/>.</param>
+        /// <param name="actualClassIndicatorIndex">The index of the actual label indicator, in the <see cref="PredictedClassesIndicators"/>.</param>
+        /// <returns></returns>
+        public double GetCountForClassPair(uint predictedClassIndicatorIndex, uint actualClassIndicatorIndex)
```

Comment:
>actualClassIndicatorIndex [](start = 83, length = 25)

Indices are usually `int`, as far as I understand. I understand that as far as they are key types they are represented as `int`, but when it comes down to, say, doing slot indexing or whatever, they're `int`s. What you think?

Reply:
Also, might be nice if there was a special property to indicate the number of classes, maybe. I guess they can get it from the above lists, but, might be nice otherwise.

---
In reply to: [275005689](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275005689) [](ancestors = 275005689)